test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/WriteJsonPayloadTests/SerializationInfoEdgeTests.cs

```diff
@@ -0,0 +1,488 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="SerializationInfoEdgeTests.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved.
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using Microsoft.AspNetCore.OData;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.E2E.TestCommon;
+using Microsoft.OData.E2E.TestCommon.Common;
+using Microsoft.OData.E2E.TestCommon.Common.Server.EndToEnd;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.Core.E2E.Tests.WriteJsonPayloadTests
+{
+    public class SerializationInfoEdgeTests : EndToEndTestBase<SerializationInfoEdgeTests.TestsStartup>
+    {
+        private readonly Uri _baseUri;
+        private readonly IEdmModel _model;
+        private static string NameSpacePrefix = "Microsoft.OData.E2E.TestCommon.Common.Server.EndToEnd.";
+
+        public class TestsStartup : TestStartupBase
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.ConfigureControllers(typeof(MetadataController));
+
+                services.AddControllers().AddOData(opt => opt.Count().Filter().Expand().Select().OrderBy().SetMaxTop(null)
+                    .AddRouteComponents("odata", CommonEndToEndEdmModel.GetEdmModel()));
+            }
+        }
+
+        public SerializationInfoEdgeTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+            _model = CommonEndToEndEdmModel.GetEdmModel();
+        }
+
+        /// <summary>
+        /// Write payloads without model or serialization info.
+        /// Note that ODL will succeed if writing requests or writing nometadata reponses.
+        /// </summary>
+        [Theory]
+        [InlineData(MimeTypeODataParameterFullMetadata)]
+        [InlineData(MimeTypeODataParameterMinimalMetadata)]
+        [InlineData(MimeTypeODataParameterNoMetadata)]
+        public async Task WriteWithoutSerializationInfo(string mimeType)

```

Comment:
It seems that this test tests different scenarios with different assertions and expectations. We create a new message writer and response for each test. It's hard to tell what scenarios are being tested from a quick glance, and it's harder to tell what the expected behaviour is because we're mixing different scenarios in the same test case. I suggest splitting the different scenarios in separate test methods, and to give each one a name that helps us understand the expected behaviour.

Reply:
I don't think this is intuitive. We are testing how to write a resource without serialization information using different mimeTypes. When using NoMimeType OData should succeed but if we are using either the MinimalMimeType or the FullMimeType then an exception should be thrown. Are you suggesting we have a test for each mime type when the expected result is the same?

Reply:
Maybe the scenarios can be put in different test cases.